### DISCUSSION TITLE
Long time between request and endpoint code get executed for large List data

### DISCUSSION THREAD

**Discussion by dfpasdel** [fastapi/fastapi#9136] [Questions] [answered]
### Description

I send image data from my client to FastApi with a request body. I noticed there is a very long time between the moment the request is sent and the moment the code in the end point starts to be executed. 

I don't need to validate input data, so I used "Any" in the base model, but the execution is slow anyway. I don't think it is related to Pydantic (cf. example). 

I joined a reproducible example, with a endpoint that only receives data but does nothing with it.

Does anyone have encountered this problem before? It might be related to the issue #1359, but my problem is on the request side and I didn't manage to extrapolate. 

It might happen also that I am addressing my objective the wrong way, so other implementation suggestions are welcomed.

```Python
from typing import List, Any
from fastapi import FastAPI
from pydantic import BaseModel
import numpy as np
import datetime
import time


class PredictRequest(BaseModel):
    #images: List[List[List[int]]]  # Even slower, overhead seems to be on the pydantic side in that case
    images: Any


app = FastAPI()

@app.post("/DummyEndPoint")
def predict(input: PredictRequest):
    print("Time data available in endpoint: ", datetime.datetime.utcnow())
    return



def run_pydantic_validation(data):
    print("PYDANTIC:")
    ts = time.time()
    validated_data = dict(PredictRequest(**data))
    print("Pydantic validation time: ", time.time() - ts)

def run_endpoint(data):
    from starlette.testclient import TestClient
    from starlette.status import HTTP_200_OK

    test_client = TestClient(app)

    ts = time.time()
    print("FASTAPI:")
    print("Time the data is sent: ", datetime.datetime.utcnow())
    response = test_client.post("/DummyEndPoint", json=data)
    assert response.status_code == HTTP_200_OK
    print("Time to run the dummy endpoint: ", time.time() - ts)


images = np.zeros((1000, 256, 256))  # Simulate 1000 images of size 256*256 (grayscale, so there is no color axis)
data = {'images': images.tolist()}

run_pydantic_validation(data=data)
run_endpoint(data=data)
```

### Environment

* OS: Linux 
* FastAPI Version: 0.61.1
* Python version: 3.7

**Comment by ArcLightSlavik** [comment]
Are you sure the images you make from numpy are correct? Running `print(images.tolist())` seemingly prints a bunch of '0' without end. 

Besides that doing `images = [[[1000, 256, 256]]]` results in good time
```
Time the data is sent: 2020-09-18 11:56:53.287025
Time data available in endpoint: 2020-09-18 11:56:53.309983
Time to run the dummy endpoint: 0.023695945739746094
```

**Comment by dfpasdel** [comment]
Thanks for your reply. 
My images from numpy are correct. I did not mentioned it clearly, but here images represents a batch of 1000 images, each being 256*256 pixels. So in the minimal example above it is a bunch of zeros, but in the real project, it is meaningful pixel values to be processed further.

**Comment by Adityagrao** [comment]
@ArcLightSlavik 
Even with setting `images = [[[1000, 256, 256]]]`  does not help, at least when I test it locally, I think the results are based on the system spec also, I am getting around 22 secs on my local machine and around 3secs on my VM

**Comment by dfpasdel** [comment, marked as answer]
I made my example even more simpler, independent from Pydantic and encoding the image data as bytes. 

```Python 
from fastapi import FastAPI, File
import numpy as np
import datetime
import time
import base64

app = FastAPI()

@app.post("/DummyEndPoint")
def predict(images: bytes = File(...)):
    print("Time data available in endpoint: ", datetime.datetime.utcnow())
    return


def run_endpoint(data):
    from starlette.testclient import TestClient
    from starlette.status import HTTP_200_OK

    test_client = TestClient(app)

    ts = time.time()
    print("FASTAPI:")
    print("Time the data is sent: ", datetime.datetime.utcnow())
    response = test_client.post("/DummyEndPoint", files=data)
    assert response.status_code == HTTP_200_OK
    print("Time to run the dummy endpoint: ", time.time() - ts)


images = np.zeros((1000, 256, 256))  # Simulate 1000 images of size 256*256 (grayscale, so there is no color axis)
data = {'images': base64.b64encode(images)}

run_endpoint(data=data)
```

The time to enter the endpoint went down on my machine from 13.5 to 2.7 sec. This time is now more reasonable with regard to the processing time in my projects endpoint. 

However I am not sure what improved here, if it is because I got rid of pydantic and/or because I converted the data to bytes (size of the data sent is almost similar, around 600MB).

**Reply by YuriiMotov** [reply]
In the first code example was passed as JSON, in this code example it's passed as "multipart/form-data".
I guess, the process of encoding\decoding is cheaper for multipart compare to JSON

**Comment by mcmx** [comment]
I'm experiencing something similar, with the TestClient, connecting to a running instance using curl or insomnia is fast enough, but the tests are taking too much, I'm using python 3.8 under pipenv, no changes on the versions of fastapi nor python under the virtualenv, one thing that happened was a change in some policies and antivirus on my Mac, will do some more testing tomorrow.

**Comment by decadance-dance** [comment]
Any updates?
